### PR TITLE
LibCompress+Utilities: Add LZMA

### PR DESCRIPTION
**Pull request by timschumi**
Pain.

This is annoyingly close to the reference implementation, at the very least the math-heavy parts that actually get into decoding instead of the high-level logic (granted, not like one can diverge much anyways if that's the only real source of "that is the formula you need to use" and "you need that many different probability counters and use them in a specific order"). The official specification is unfortunately basically just the reference implementation with a bit of text around the logic parts, and I wasn't able to find any human-language specifications that have actually been completed.

I hope that this is still acceptable, considering that the specification/reference implementation has been placed in the public domain.

## Remaining TODOs/discussion points

 - ~~Fuzzer: Not viable at the moment, since the archive can legitimately request up to 4GB of lookback buffer size, which instantly OOMs LibFuzzer.~~
 - ~~Specification quotes: The implementation currently features a mix of quotes from the specification and some comments of my own. Suggestions on how to make those visually distinct would be very welcome, if that is a priority.~~
 - ~~Tests: Will add tomorrow.~~
 - ~~Rough edges: Nothing is explicitly unimplemented at the moment, but there are some noticeable edge cases that I haven't looked into yet (for example, some of the testing files from the SDK don't even load properly).~~

### CODE REVIEW

File: Userland/Libraries/LibCompress/Lzma.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    u8 literal_context_bits;
+    u8 literal_position_bits;
+    u8 position_bits;
+    u32 dictionary_size;
+    Optional<u64> uncompressed_size;
+};
+
+class LzmaDecompressor : public Stream {
+public:
+    static ErrorOr<NonnullOwnPtr<LzmaDecompressor>> create(NonnullOwnPtr<Stream>, LzmaDecompressorOptions const&);
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibCompress/Lzma.h:29]: Should this not take `Handle<Stream>`?

File: Userland/Libraries/LibArchive/Lzma.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+private:
+    u8 m_model_properties;
+    u32 m_dictionary_size;
+    u64 m_uncompressed_size;
+};
+static_assert(sizeof(LzmaHeader) == 13);
+
+class Lzma {
+public:
+    static ErrorOr<NonnullOwnPtr<Compress::LzmaDecompressor>> create(NonnullOwnPtr<Stream>);
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibArchive/Lzma.h:35]: Same here with Handle

File: Userland/Utilities/lzcat.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <LibMain/Main.h>
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    StringView filename;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(filename, "File to decompress", "file");
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/lzcat.cpp:17]: Should probably have a short help text like "Decompress and print an LZMA archive"

File: Userland/Utilities/lzcat.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    args_parser.parse(arguments);
+
+    NonnullOwnPtr<Stream> stream = TRY(Core::File::open(filename, Core::File::OpenMode::Read));
+    stream = TRY(Archive::Lzma::create(move(stream)));
+
+    // Arbitrarily chosen buffer size.
+    Array<u8, 4096> buffer;
```
Comment (kleinesfilmroellchen) [inline Userland/Utilities/lzcat.cpp:24]: Use PAGE_SIZE?

File: Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+#include <AK/MemoryStream.h>
+#include <LibCompress/Lzma.h>
+#include <stdio.h>
+
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* data, size_t size)
+{
+    // These settings have been taken from the a.lzma example in the LZMA specification.
+    // Fuzzing the settings is not currently viable as archives can request a dictionary size of up to 4GB.
```
Comment (kleinesfilmroellchen) [inline Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:14]: Can we fuzz all options *except* the dictionary size? i.e. overwrite the dictionary size option after reading the options from the provided data.

File: Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        .uncompressed_size = {},
+    };
+
+    auto stream = make<FixedMemoryStream>(ReadonlyBytes { data, size });
+    auto decompressor_or_error = Compress::LzmaDecompressor::create(move(stream), options);
+    if (decompressor_or_error.is_error())
+        return 0;
+    auto decompressor = decompressor_or_error.release_value();
+    while (!decompressor->is_eof()) {
+        auto maybe_error = decompressor->discard(4096);
```
Comment (kleinesfilmroellchen) [inline Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:29]: PAGE_SIZE as well probably

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibCompress/Lzma.h (hunk 1 above)
Comment (timschumi) [reply Userland/Libraries/LibCompress/Lzma.h:29]: It's called `MaybeOwned` now, but yes, it should (at the very least for LZMA2). :^)

File: Userland/Utilities/lzcat.cpp (hunk 4 above)
Comment (timschumi) [reply Userland/Utilities/lzcat.cpp:24]: `PAGE_SIZE` is not consteval on Lagom, so we can't use it here. It also isn't nearly as important to align to pages compared to heap allocations. In this case, it really is just a nice and round number.

File: Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp (hunk 5 above)
Comment (timschumi) [reply Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp:14]: We could do that, but that would result in files that look like LZMA archives (not that they have some kind of magic), but would have to be changed first to yield the same results as the fuzzer.

What we _could_ do is to define some arbitrary "largest reasonable dictionary size" and reject any test inputs that have a larger dictionary size than that. LibFuzzer has a dedicated return value to indicate "don't add this to the test corpus", which would be perfect for this.

Comment (kleinesfilmroellchen) [reply Meta/Lagom/Fuzzers/FuzzLzmaDecompression.cpp]: That sounds like the ideal solution to me then.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by timschumi** [comment]
Renamed `Archive::Lzma` to `Archive::LzmaDecompressor`, to match `Compress::LzmaDecompressor`. But now this should really be ready for merging. :^)

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by GMTA** [comment]
> Specification quotes: The implementation currently features a mix of quotes from the specification and some comments of my own. Suggestions on how to make those visually distinct would be very welcome, if that is a priority.

I try to explicitly reference section numbers when possible, and literally quote (`"The spec says that..."`) the parts that I copy over from the spec.

**Comment by timschumi** [comment]
I'm afraid the specification is a bit short on section numbers, and it uses section names/headers _very_ sparingly. I will try adding quotes around the quotes and prefixing them with some kind of distinctive text to see how it looks in practice.

**Comment by timschumi** [comment]
Changes in latest push:
 - Rebased onto and adapted to the latest Stream function name changes.
 - Removed some unnecessary braces around one line for-loops.
 - Visually marked verbatim specification comments by wrapping them in `"`.
 - Replaced the specification vs. reference implementation FIXME with a "Note", since I experimentally confirmed that the reference implementation is most likely correct.

**Review by kleinesfilmroellchen — APPROVED** [review]

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 7]
```diff
@@ -0,0 +1,604 @@
...
+    constexpr size_t probability_shift_width = 5;
+
+    u32 bound = (m_range_decoder_range >> probability_bit_count) * probability;
+
+    if (m_range_decoder_code < bound) {
+        probability += ((1 << probability_bit_count) - probability) >> probability_shift_width;
+        m_range_decoder_range = bound;
+        TRY(normalize_range_decoder());
+        return 0;
+    } else {
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:139]: no need for else after return

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+
+    return result;
+}
+
+ErrorOr<void> LzmaDecompressor::decode_literal_to_output_buffer()
+{
+    u8 previous_byte = 0;
+    if (m_total_decoded_bytes > 0) {
+        auto read_bytes = MUST(m_output_buffer.read_with_seekback({ &previous_byte, sizeof(previous_byte) }, 1));
+        VERIFY(read_bytes.size() == sizeof(previous_byte));
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:190]: Not sure I understand this verify, if `read_with_seekback` didn't return an error `read_bytes` is a guarenteed to return 1 byte right? Otherwise it's a bug in the `read_with_seekback` implementation? If so, this seems a bit excessive, we should either assert this inside the implementation itself, or not at all.

Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:190]: (same for everywhere else below)

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+    }
+
+    // "To select the table for decoding it uses the context that consists of
+    //  (lc) high bits from previous literal and (lp) low bits from value that
+    //  represents current position in outputStream."
+    u16 literal_state_bits_from_position = m_total_decoded_bytes & ((1 << m_options.literal_position_bits) - 1);
+    u16 literal_state_bits_from_output = previous_byte >> (8 - m_options.literal_context_bits);
+    u16 literal_state = literal_state_bits_from_position << m_options.literal_context_bits | literal_state_bits_from_output;
+
+    Span<Probability> selected_probability_table = m_literal_probabilities.span().slice(0x300 * literal_state, 0x300);
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:200]: We use this `0x300` constant in multiple places, maybe make it a constexpr constant?

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 10]
```diff
@@ -0,0 +1,604 @@
...
+
+    size_t written_bytes = m_output_buffer.write({ &actual_result, sizeof(actual_result) });
+    VERIFY(written_bytes == sizeof(actual_result));
+    m_total_decoded_bytes += sizeof(actual_result);
+
+    return {};
+}
+
+LzmaDecompressor::LzmaLengthDecoderState::LzmaLengthDecoderState()
+{
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:242]: ditto regarding the helper from above

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 11]
```diff
@@ -0,0 +1,604 @@
...
+}
+
+bool LzmaDecompressor::is_eof() const
+{
+    if (m_output_buffer.used_space() > 0)
+        return false;
+
+    if (m_options.uncompressed_size.has_value()) {
+        return m_total_decoded_bytes >= m_options.uncompressed_size.value();
+    }
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:590]: nitpick, unnecessary curlies

File: Userland/Libraries/LibArchive/Lzma.cpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    //  Encoded using `((pb * 5 + lp) * 9 + lc)`."
+
+    u8 input_bits = m_model_properties;
+
+    if (input_bits >= (9 * 5 * 5))
+        return Error::from_string_literal("Encoded model properties value is larger than the highest possible value");
+
+    literal_context_bits = input_bits % 9;
+    input_bits /= 9;
+    VERIFY(literal_context_bits >= 0 && literal_context_bits <= 8);
```
Comment (ADKaster) [inline Userland/Libraries/LibArchive/Lzma.cpp:60]: Just to confirm: These properties on the bits are mandated by either the math here in this function or by the caller of this function? (i.e. we won't see any fuzzer crashes hitting these asserts, right?)

File: Userland/Libraries/LibArchive/Lzma.h [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+private:
+    u8 m_model_properties;
+    u32 m_dictionary_size;
+    u64 m_uncompressed_size;
+};
+static_assert(sizeof(LzmaHeader) == 13);
+
+class LzmaDecompressor {
+public:
+    static ErrorOr<NonnullOwnPtr<Compress::LzmaDecompressor>> create(MaybeOwned<Stream>);
```
Comment (ADKaster) [inline Userland/Libraries/LibArchive/Lzma.h:37]: Archive::LzmaDecompressor::create() returns a Compress::LzmaDecompressor? This seems kind of odd, but then I didn't cross reference it with the other Decompressors that mostly just redirect to LibCompress.

File: Userland/Utilities/lzcat.cpp [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    StringView filename;
+
+    Core::ArgsParser args_parser;
+    args_parser.set_general_help("Decompress and print an LZMA archive");
+    args_parser.add_positional_argument(filename, "File to decompress", "file");
+    args_parser.parse(arguments);
+
+    NonnullOwnPtr<Stream> stream = TRY(Core::File::open(filename, Core::File::OpenMode::Read));
```
Comment (ADKaster) [inline Userland/Utilities/lzcat.cpp:21]: Thoughts on using that "open file or standard stream" thing here? or does Posix/Linux lzcat not support `-` as an arg.

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 15]
```diff
@@ -0,0 +1,604 @@
...
+        // "The decoder calculates "state2" variable value to select exact variable from
+        //  "IsMatch" and "IsRep0Long" arrays."
+        u16 position_state = m_total_decoded_bytes & ((1 << m_options.position_bits) - 1);
+        u16 state2 = (m_state << maximum_number_of_position_bits) + position_state;
+
+        auto update_state_after_literal = [&] {
+            if (m_state < 4)
+                m_state = 0;
+            else if (m_state < 10)
+                m_state = m_state - 3;
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:379]: nitpick, but `-=`

**Review by ADKaster — COMMENTED** [review]
Couple thoughts for clarification, but looks good :^)

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 16]
```diff
@@ -0,0 +1,604 @@
...
+        // "The decoder uses the following code flow scheme to select exact
+        //  type of LITERAL or MATCH:
+        //
+        //  IsMatch[state2] decode
+        //   0 - the Literal"
+        if (TRY(decode_bit_with_probability(m_is_match_probabilities[state2])) == 0) {
+            // "At first the LZMA decoder must check that it doesn't exceed
+            //  specified uncompressed size."
+            if (m_options.uncompressed_size.has_value() && m_options.uncompressed_size.value() <= m_total_decoded_bytes)
+                return Error::from_string_literal("Exceeded uncompressed archive size while decoding literal symbol");
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:446]: we already check this above, so this is unrechable. It would only be reachable if we did another iteration of the loop to check for the EOF marker, which we don't.

Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:446]: (I would keep the spec text, but add a comment explaining it's not unecessary, assuming you agree with me of course :slightly_smiling_face:)

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 17]
```diff
@@ -0,0 +1,604 @@
...
+
+            // "If the value of "rep0" is equal to 0xFFFFFFFF, it means that we have
+            //  "End of stream" marker, so we can stop decoding and check finishing
+            //  condition in Range Decoder"
+            if (m_rep0 == 0xFFFFFFFF) {
+                if (m_range_decoder_code != 0)
+                    return Error::from_string_literal("Found 'end of stream' marker, but range decoder is in unfinished state");
+
+                m_found_end_of_stream_marker = true;
+                continue;
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:482]: We set `m_found_end_of_stream_marker` (this is the only place where we do so), and immediately continue the loop, at which point we break out of it using the loop end condition at the top. We then go into the post-loop check for the code validity if we reached the end of stream marker.
AKA: This check for the final code value is duplicated, we should remove one of the copies.

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 18]
```diff
@@ -0,0 +1,604 @@
...
+            }
+
+            // "If uncompressed size is defined, LZMA decoder must check that it doesn't
+            //  exceed that specified uncompressed size."
+            if (m_options.uncompressed_size.has_value() && m_options.uncompressed_size.value() <= m_total_decoded_bytes)
+                return Error::from_string_literal("Exceeded uncompressed archive size while decoding simple match symbol");
```
Comment (IdanHo) [inline Userland/Libraries/LibCompress/Lzma.cpp:488]: same as above

**Review by IdanHo — COMMENTED** [review]
nice work, mostly nitpicks

File: Userland/Libraries/LibArchive/Lzma.h (hunk 13 above)
Comment (timschumi) [reply Userland/Libraries/LibArchive/Lzma.h:37]: The reason why I made this separate is because LZMA technically defines an 'archive' format (which is more of a stream container, really).

It is a really puny archive format, because all it does is prefixing it with a small header that notes down the decoder settings for the contained stream, but it is distinct from the other LibCompress-supported algorithms that really only dump the raw stream into a file and call it a day.

The reason why it returns a `Compress::LzmaDecompressor` is because we read the header upon creation of the stream, configure the underlying `LibCompress` stream once, and would then just delegate all the time anyways. Returning the underlying Stream directly and not implementing the whole Stream interface for Archive::LzmaDecompressor again seemed like a better idea.

Would we rather want to have a `Compress::LzmaDecompressor::create_from_archive` (/container?) and `Compress::LzmaDecompressor::create_from_raw_stream`, where one relies on the other?

File: Userland/Libraries/LibArchive/Lzma.h [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+};
```
[the remark this answers is not in the document]
Comment (timschumi) [reply Userland/Libraries/LibArchive/Lzma.h:38]: The question regarding the return type I have started to discuss in a response to Andrew, so we should probably continue to do that there exclusively.

File: Userland/Libraries/LibArchive/Lzma.h (hunk 13 above)
Comment (timschumi) [reply Userland/Libraries/LibArchive/Lzma.h:37]: From Idans comment about the same topic:

> The distiniction between a compression format and an archival format, is that a compression format takes a stream of bytes and compresses/decompresses it, while an archive format packs together one or more files, with the file contents possibly being compressed using a compression format.

Ok, in that case LZMA 'archives' wouldn't be an actual archive because (in comparison to an LZMA compressed stream) they only provide the settings for the singular stream that they contain.

However, looking at what's to come, are LZMA2 and XZ archive formats then? Should they also rather be in LibCompress?

Neither of them have a concept of one or multiple files, both just encode a stream of data, but this time in multiple ways. LZMA2 splits the stream in parts and either stores them LZMA-compressed or uncompressed. XZ allows compressing its data stream using LZMA2 and/or a bunch of of other so-called filters. None of them know what 'files' are, if a user needs that they have to TAR it up first.

File: Tests/LibCompress/TestLzma.cpp [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+        0x5A, 0x9E, 0x10, 0x3A, 0x4F, 0x65, 0xFA, 0x41, 0xCB, 0xF2, 0x79, 0x65, 0xD7, 0xF1, 0x9F, 0xAB,
+        0x70, 0x1D, 0x6F, 0xF7, 0xB6, 0x79, 0xCC, 0x8A, 0x7D, 0xCE, 0xDB, 0xF8, 0xF6, 0x9E, 0xC9, 0x12,
+        0x9F, 0xAA, 0xBF, 0x89, 0xFE, 0x05, 0x36, 0x80
+    };
+
+    auto stream = MUST(try_make<FixedMemoryStream>(compressed));
+    auto decompressor = MUST(Compress::LzmaDecompressor::create(move(stream), options));
+    // FIXME: This should detect that there is still remaining (non-'end of stream') data after the expected data and reject the last read.
+    //        As of now, this test accepts any result, except for crashes.
+    (void)decompressor->read_until_eof(PAGE_SIZE);
```
[the remark this answers is not in the document]
Comment (timschumi) [reply Tests/LibCompress/TestLzma.cpp:213]: > Are we going to fix this FIXME?

I plan to, yes. However, since my first attempt at this led to a bunch of erroneous failures and end-of-stream detection will have a significant effect on formats that embed LZMA, I decided to let it be for now and revisit it later, focusing on getting the main parts of the decoder (and the followup support for LZMA2 and XZ) finished up and merged first.

> Also, we might want to expose a is_corrupted helper that returns true if we reached the 'known' uncompressed size but m_range_decoder_code is non-0? The spec says that's a good indicator for an incorrectly decoded stream (and you already check that when we reach the explicit end indicator), so we can assert that returns true here.

The specification indeed mentions that this is a good way of checking for corruption, but it unfortunately doesn't seem to hold up if the archive contains an uncompressed size _and_ an end of stream marker (since it still has to read the end marker at that point). This is also what I suspect was causing my failures that shouldn't have been there.

In any case, the current mode of operation of this implementation is to stop and report an error even for things that would otherwise just be "corrupted", so I'd rather have it report that immediately if it even just notices that something is wrong.

**Review by timschumi — COMMENTED** [review]
Fixed a large part of the nitpicks, left comments on the others.

File: Userland/Libraries/LibArchive/Lzma.h (hunk 13 above)
Comment (IdanHo) [reply Userland/Libraries/LibArchive/Lzma.h:37]: Comparing to the Gzip, which we already have implemented in serenity, it seems to match up: It's based on Deflate, it adds a short header at the start with the compression options, and is possibly used to compress tar archives.
Currently the implementation for Gzip is in LibCompress, matching my expectations at least.

The only unfortunate part about LZMA is that the name is used both for the raw stream and the stream with the header part.

I would say all 3 of LZMA, LZMA2 and XZ should end up in LibCompress.

File: Tests/LibCompress/TestLzma.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (IdanHo) [reply Tests/LibCompress/TestLzma.cpp:213]: I see, let's leave that to a future PR then :+1:

File: Userland/Libraries/LibCompress/Lzma.h [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    public:
+        LzmaLengthDecoderState();
+
+        Probability m_first_choice_probability { default_probability };
+        Probability m_second_choice_probability { default_probability };
+
+        static constexpr size_t maximum_number_of_position_bits = 4;
+        Array<Array<Probability, (1 << 3)>, (1 << maximum_number_of_position_bits)> m_low_length_probabilities;
+        Array<Array<Probability, (1 << 3)>, (1 << maximum_number_of_position_bits)> m_medium_length_probabilities;
+        Array<Probability, (1 << 8)> m_high_length_probabilities;
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.h:83]: Ah, if it's just the chosen value per tree then it's fine, I thought `1 << 3` was being duplicated, but it sounds like it's actually 2 different values, low tree length, and medium tree length, which just ended up as the same value.

File: Userland/Libraries/LibCompress/Lzma.cpp [hunk 22]
```diff
@@ -0,0 +1,604 @@
...
+}
+
+ErrorOr<u8> LzmaDecompressor::decode_direct_bit()
+{
+    m_range_decoder_range >>= 1;
+    m_range_decoder_code -= m_range_decoder_range;
+
+    // The specification (or rather, the reference implementation) does not specify what "t" is. As such, the name has been carried over unchanged.
+    // Since the human language description on Wikipedia decides the return value by comparing range and code, "t" appears to be a simple helper for a branchless implementation.
+    u32 t = 0 - (m_range_decoder_code >> 31);
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:96]: Yep, I prefer so, simply to make the code a tiny bit more readable.

File: Userland/Libraries/LibCompress/Lzma.cpp (hunk 7 above)
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:124]: nitpicks can be ignored when reasonable :slightly_smiling_face:

File: Userland/Libraries/LibCompress/Lzma.cpp (hunk 9 above)
Comment (IdanHo) [reply Userland/Libraries/LibCompress/Lzma.cpp:185]: I think `probability_table_size` is fine

**Review by IdanHo — APPROVED** [review]
LGTM, thanks! Support for new compression formats is always exciting :^)
